SR-7266: Avoid nesting of ReversedCollection

Moximillian opened the pull request:
Applying reversed() two times should return original collection.

<!-- What's in this pull request? -->
First round of fixes for the SR-7266, to avoid nesting multiple ReversedCollections and instead return the original collection with two consecutive .reversed() functions.

Does not yet fix the .lazy use case.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-7266](https://bugs.swift.org/browse/SR-7266).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

danielctull committed:
[stdlib] SR-7266 Add ReverseCompatibility test

Moximillian committed:
SR-7266: Avoid nesting of ReversedCollection

Applying reversed() two times should return identity.

danielctull commented:
Hi there, I started looking at this bug last weekend with a solution to the lazy scenario, which you can find at this gist: https://gist.github.com/danielctull/ec521f3fc2aadabb17d38f716e55c5f7

It seems tricky to constrain `LazyCollectionProtocol` in a way to overload a version of the `reversed()` function when the lazy's base is itself a `ReversedCollection`. I've been meaning to raise this as a PR, to see whether the additional protocol would justified for use only in this scenario.

Moximillian commented:
Hi, yes, I also ended up with having to create ReversedCollectionProtocol for the lazy case. It's now updated to the PR.

Moximillian commented:
As a side question to the reviewers, is there a reason why these internal variables are public?
LazySequence.swift:  public var _base: Base
Reverse.swift:             public let _base: Base

moiseev commented on stdlib/public/core/Reverse.swift:
```diff
@@ -253,6 +253,19 @@ extension ReversedCollection: BidirectionalCollection {
 extension ReversedCollection: RandomAccessCollection where Base: RandomAccessCollection { }
 
+extension ReversedCollection {
+  /// This is optimization to return identity of doubly reversed collection
```
This part of the comment is an implementation detail, I don't think it should be a doc-comment.

moiseev reviewed (commented):
@Moximillian, thanks for addressing the issue!

AFAIK @danielctull has tests and properly set availability attributes for the old and new `reversed` overloads. It would make a lot of sense for you two to merge efforts.

Let's concentrate on a non-lazy case first.

moiseev commented on stdlib/public/core/Reverse.swift:
```diff
@@ -253,6 +271,16 @@ extension ReversedCollection: BidirectionalCollection {
+  // For example [1,2].reversed().reversed() => [1,2]
+
+  /// Returns the elements of the original collection (reversal of reversal)
+  public func reversed() -> Base {
```
Since this is a more concrete overload of an existing method, we should be careful not to break existing code.

The type of `let xs = [1,2,3].reversed().reversed()` changes. This new overload should be `@available(swift, introduced: 4.2)`.

danielctull commented:
As @moiseev said, I have a branch which adds tests and availability attributes which you can see in this [comparison of master:SR-7266](https://github.com/apple/swift/compare/master...danielctull-forks:SR-7266).

Moximillian commented:
My first starterbug so please bear with me :)

@moiseev:
1) I amended the comments to align with @danielctull 
2) added availability and inline
3) added the testcase for non-lazy variant from @danielctull 
3) I don't know if you meant this, but the PR now only contains the non-lazy case, which means it does not add or use the new ReversedCollectionProtocol at all.

danielctull commented:
Hi @Moximillian! I've sent you a more long winded email but to paraphrase here: Because I'm a little vain and I'd like to be able to have recognition for my work, I split my test commit into the regular and lazy checks and rebased that into your branch. You can see this here: https://github.com/danielctull-forks/swift/commits/SR-7266-avoid-nesting-reversed. It keeps all of you other changes, but allows me to be the committer for the test additions.

You could either reset my branch or cherry-pick my commit [dbdd74a8a8fcf6f70cc9810fe81b5efaa18bd57e](https://github.com/danielctull-forks/swift/commit/dbdd74a8a8fcf6f70cc9810fe81b5efaa18bd57e) and force push to your fork to update this PR.

danielctull committed:
[stdlib] SR-7266 Make sure to actually run the tests

danielctull commented:
Cool, cheers @Moximillian. I think this line to run the test was lost when I split out the lazy tests earlier, if you want to merge this one in too: [110655befd06952c82f4f28406476da573962d8c](https://github.com/danielctull-forks/swift/commit/110655befd06952c82f4f28406476da573962d8c)

I ran these on the previous branch (before you squashed) and it ran fine! 😁

Moximillian commented:
Cherry-picked the missing line as well

moiseev commented:
@swift-ci Please test

moiseev commented:
@swift-ci Please Test Source Compatibility

moiseev commented:
@swift-ci Please test

moiseev commented:
@swift-ci Please Test Source Compatibility

moiseev commented on stdlib/public/core/Reverse.swift:
```diff
@@ -84,7 +82,7 @@ extension ReversedCollection {
   }
 }
- 
+
```
Oh. I misinterpreted the diff. I would prefer to not mix the cleanups with functional changes, but am guilty of doing it every now and then myself. So let's keep it.

moiseev commented:
For the record. The binary size difference is negligible.
